modules/javafx.graphics/src/main/native-glass/gtk/glass_dnd.cpp

```diff
@@ -760,132 +734,346 @@ static gboolean dnd_source_set_raw(GtkWidget *widget, GtkSelectionData *sel_data
-        dnd_source_set_string(widget, data, atom);
-    } else if (target_is_image(atom)) {
-        dnd_source_set_image(widget, data, atom);
-    } else if (target_is_uri(atom)) {
-        dnd_source_set_uri(widget, data, atom);
+static void process_drag_motion(DragSourceContext *ctx, gint x_root, gint y_root, guint state) {
+    GdkWindow *dest_window;
+    GdkDragProtocol prot;
+    GdkWindow *ignore = NULL;
+
+    ctx->last_x = x_root;
+    ctx->last_y = y_root;
+
+    if (ctx->drag_view) {
+        ctx->drag_view->move(x_root, y_root);
+        ignore = ctx->drag_view->get_window();
+    }
+
+    gdk_drag_find_window_for_screen(ctx->dnd_ctx, ignore, gdk_screen_get_default(),
+            x_root, y_root, &dest_window, &prot);
+
+    ctx->dest_window = dest_window;
+
+    if (prot != GDK_DRAG_PROTO_NONE) {
+        GdkDragAction action, possible_actions;
+        determine_actions(ctx, state, &action, &possible_actions);
+        gdk_drag_motion(ctx->dnd_ctx, dest_window, prot, x_root, y_root,
+                action, possible_actions, GDK_CURRENT_TIME);
+    }
+}
+
+static void process_dnd_source_mouse_motion(DragSourceContext *ctx, GdkEvent *event) {
+    GdkEventMotion *eventMotion = &event->motion;
+    process_drag_motion(ctx, eventMotion->x_root, eventMotion->y_root, eventMotion->state);
+}
+
+static void process_dnd_source_key_press_release(DragSourceContext *ctx, GdkEvent *event) {
+    GdkEventKey *eventKey = &event->key;
+
+    if (eventKey->is_modifier) {
+        guint state = eventKey->state;
+        guint new_mod = 0;
+        gint x,y;
+        if (eventKey->keyval == GLASS_GDK_KEY_CONSTANT(Control_L) ||
+                eventKey->keyval == GLASS_GDK_KEY_CONSTANT(Control_R)) {
+            new_mod = GDK_CONTROL_MASK;
+        } else if (eventKey->keyval == GLASS_GDK_KEY_CONSTANT(Alt_L) ||
+                eventKey->keyval == GLASS_GDK_KEY_CONSTANT(Alt_R)) {
+            new_mod = GDK_MOD1_MASK;
+        } else if (eventKey->keyval == GLASS_GDK_KEY_CONSTANT(Shift_L) ||
+                eventKey->keyval == GLASS_GDK_KEY_CONSTANT(Shift_R)) {
+            new_mod = GDK_SHIFT_MASK;
+        }
+
+        if (eventKey->type == GDK_KEY_PRESS) {
+            state |= new_mod;
+        } else {
+            state ^= new_mod;
+        }
+
+        process_drag_motion(ctx, ctx->last_x, ctx->last_y, state);
+    }
+}
+
+static GdkCursor * get_action_cursor(GdkDragAction selected) {
+    GdkCursor* cursor = NULL;
+
+    if (selected & GDK_ACTION_COPY) {
+        cursor = gdk_cursor_new_from_name(gdk_display_get_default(), "dnd-copy");
+    } else if (selected & (GDK_ACTION_MOVE | GDK_ACTION_PRIVATE)) {
+        cursor = gdk_cursor_new_from_name(gdk_display_get_default(), "dnd-move");
+    } else if (selected & GDK_ACTION_LINK) {
+        cursor = gdk_cursor_new_from_name(gdk_display_get_default(), "dnd-link");
+    }
+
+    if (cursor == NULL) {
+        cursor = gdk_cursor_new_from_name(gdk_display_get_default(), "dnd-none");
+    }
```

Comment:
Those cursor works on Ubuntu 16.04+ (probably before) and are consistent with the ones used on gnome.